internal/config/config.go

```diff
@@ -556,11 +643,59 @@ func validInstanceTypeForProvider(insType string, acceptNonCVM bool, provider cl
 			}
 		}
 		return false
+	case cloudprovider.GCP:
+		for _, instanceType := range instancetypes.GCPInstanceTypes {
+			if insType == instanceType {
+				return true
+			}
+		}
+		return false
 	default:
 		return false
 	}
 }
 
+// checkIfAWSInstanceTypeIsValid checks if an AWS instance type passed as user input is in one of the instance families supporting NitroTPM-
+func checkIfAWSInstanceTypeIsValid(userInput string) bool {
```

Comment:
Some targeted unit tests for this func would be great. Is the AWS instance type format documented somewhere officially? I actually don't know and do not have experience with AWS instance types. If we do custom parsing (instead of listing all available types which - granted - might be a lot), we should ensure that there are no false positives / negatives.

Reply:
TestValidInstanceTypeForProvider essentially calls this function directly. I could create a targeted test but it would essentially be a copy-paste (or removal of the AWS test cases from this function). Choose what you prefer.

The general naming scheme is described here:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html#AvailableInstanceTypes

Of course in all cleverness Amazon does not always follow the convention with 100% accuracy. For example, there's this whole family of Graviton instances called the A1 family, and none of them has `g` in their name:
https://aws.amazon.com/de/ec2/instance-types/a1/

A for ARM, a for AMD, g for Graviton (which is ARM but okay), _right?_ 🙄 

But they are not on the supported list anyways, luckily. 

I think a better explanation can be found here:
https://stackoverflow.com/a/56880093

On that note, I maybe should switch to M6a as the default value.

From what I can tell, xlarge is almost always 4vCPUs, and everything beyond that are multiplicators of xlarge or metal (which NitroTPM does not support, according to the docs). I haven't seen an exception to this yet, luckily. So everything needs to be xlarge or higher (but not metal!) to account for the 2:1 vCPU ratio.

And from what I can tell, all supported instances are indeed following the 2:1 vCPU ratio:
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/cpu-options-supported-instances-values.html

So at least with the current set of supported instance types I don't see where the parser can cause issues since family parsing and size parsing should be mostly okay I believe, and the modificators are part of the family. I need to test case-sensitive quickly, though.

Small downside of the approach right now is that we don't support checking the specific instance type. For example, if someone writes "x2gd.7xlarge" we don't check for that and it will fail at creation. But personally I am not sure if we want to check every size. It's definitely possible, but it's too many families to keep this list up-to-date I guess.

Reply:
I see. I think we can leave it as-is for now :) Thanks for the pointers to your research.